bin/src/hickory-dns.rs

```diff
@@ -925,3 +929,70 @@ fn build_udp_socket(ip: IpAddr, port: u16) -> Result<UdpSocket, Error> {
 
     UdpSocket::from_std(sock.into())
 }
+
+/// Drop privileges on Unix systems if running as root. Errors that prevent dropping privileges will
+/// halt the server.  This must be called after binding to low numbered sockets is complete.
+#[cfg(target_family = "unix")]
+fn check_drop_privs(user: &str, group: &str) {
+    use libc::{getegid, geteuid, getgid, getgrnam, getpwnam, getuid, setgid, setuid};
+    use std::ffi::CString;
+
+    let (uid, gid, euid, egid) = unsafe { (getuid(), getgid(), geteuid(), getegid()) };
+
+    if uid == 0 || euid == 0 {
+        let Ok(user_cstring) = CString::new(user) else {
+            error!("unable to create CString for user {user}. Exiting.");
+            process::exit(1);
+        };
+
+        let Ok(group_cstring) = CString::new(group) else {
+            error!("unable to create CString for group {group}. Exiting.");
+            process::exit(1);
+        };
+
+        info!(
+            "running as root (uid: {uid} gid: {gid} euid: {euid} egid: {egid})...dropping privileges.",
+        );
+
+        let (user_info, group_info) = unsafe {
+            (
+                getpwnam(user_cstring.as_ptr()),
+                getgrnam(group_cstring.as_ptr()),
+            )
+        };
+
+        if user_info.is_null() {
+            error!("unable to lookup user '{user}'. Exiting.");
+            process::exit(1);
+        }
+
+        if group_info.is_null() {
+            error!("unable to lookup group '{group}'. Exiting.");
+            process::exit(1);
+        }
+
+        let (setgid_rc, setuid_rc) =
+            unsafe { (setgid((*group_info).gr_gid), setuid((*user_info).pw_uid)) };
+
+        if setgid_rc < 0 {
+            error!("unable to set gid. Exiting.");
+            process::exit(1);
+        }
+
+        if setuid_rc < 0 {
+            error!("unable to set uid. Exiting.");
+            process::exit(1);
+        }
+    } else {
+        info!("hickory did not start as root, not dropping privileges");
+    }
```

Comment:
This feels superfluous, especially when you're going to report the context later anyway.